pkg/store/http.go

```diff
@@ -0,0 +1,355 @@
+		return nil, fmt.Errorf("error constructing hash.Writer: %w", err)
+	}
+
+	// setup a io.TeeReader to ensure the nar file contents get written to hashWriter
+	// while the NarReader reads through the body.
+	tr := io.TeeReader(narReq.Body, hashWriter)
+
+	// receive the .nar file
+	pathInfo, err := Import(ctx, ni.StorePath, ni.References, tr, hs.cacheChunkStore)
+	if err != nil {
+		return nil, fmt.Errorf("error converting narinfo and nar to pathInfo: %w", err)
+	}
+
+	// query the hashWriter if size matches.
+	if ni.NarSize != hashWriter.BytesWritten() {
+		return nil, fmt.Errorf("read %v bytes of nar, expected %v", ni.NarSize, hashWriter.BytesWritten())
+	}
+
+	// query the hashWriter if hash matches
+	if !bytes.Equal(ni.NarHash.Digest, hashWriter.Digest()) {
+		return nil, fmt.Errorf("got %s:%s as NarHash, expected %s",
+			ni.NarHash.HashType,
+			nixbase32.EncodeToString(hashWriter.Digest()),
+			ni.NarHash,
+		)
+	}
+
+	return pathInfo, nil
+}
+
+func (hs *HTTPStore) Has(ctx context.Context, outputPath string) (bool, error) {
+	np, err := nixpath.FromString(outputPath)
+	if err != nil {
+		return false, fmt.Errorf("error parsing output path: %w", err)
+	}
+
+	// check the underlying cache store first
+	has, err := hs.cacheStore.Has(ctx, outputPath)
+	if err != nil {
+		return false, fmt.Errorf("error asking underlying cache store: %w", err)
+	}
+	// if it's in there, we can return true.
+	if has {
+		return true, nil
+	}
+
+	// else, we peek at the .narinfo file with a HEAD request.
+	niURL := hs.getNarinfoURL(np)
+
+	// construct the request
+	niReq, err := http.NewRequestWithContext(ctx, "HEAD", niURL.String(), nil)
+	if err != nil {
+		return false, fmt.Errorf("error constructing request: %w", err)
+	}
+
+	// do the request for the .narinfo file
+	niResp, err := hs.Client.Do(niReq)
+	if err != nil {
+		return false, fmt.Errorf("error doing narinfo request: %w", err)
+	}
+	defer niResp.Body.Close()
+
+	// if we get a 404, we assume it doesn't exist.
+	if niResp.StatusCode == http.StatusNotFound {
+		return false, nil
+	}
+
+	// return an error in case other errors occur
+	// TODO: check if there's other status codes that should return false, nil
+	if niResp.StatusCode < 200 || niResp.StatusCode >= 300 {
+		return false, fmt.Errorf("bad status code retrieving narinfo: %v", niResp.StatusCode)
+	}
+
+	// else, return true.
+	return true, nil
+}
+
+func (hs *HTTPStore) Put(ctx context.Context, pathInfo *PathInfo) error {
+	np, err := nixpath.FromString(pathInfo.OutputName)
+	if err != nil {
+		return fmt.Errorf("error parsing output path: %w", err)
+	}
+
+	// Usually NAR files are uploaded to /nar/$narhash.nar[.$compressionSuffix].
+	// However, this means we need to know the NARHash before uploading the files, as
+	// a plain HTTP PUT call contains the destination path, and we can't move files after upload.
+
+	// This means, we render the NAR file twice - once to calculate NarHash, NarSize,
+	// a second time to do the actual upload.
```

Comment:
Maybe that's an argument for actually storing `NarHash` and `NarSize` in the `PathInfo`.

@adisbladis, what do you think?